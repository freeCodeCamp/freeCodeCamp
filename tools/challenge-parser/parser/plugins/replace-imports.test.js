const path = require('path');
const cloneDeep = require('lodash/cloneDeep');
const toVfile = require('to-vfile');
const selectAll = require('unist-util-select').selectAll;
const parseFixture = require('../__fixtures__/parse-fixture');

const addImports = require('./replace-imports');

describe('replace-imports', () => {
  let importsAST,
    importsTwoAST,
    importsExtraAST,
    simpleAST,
    markerAST,
    correctFile,
    incorrectFile;

  let originalImportsAST,
    originalImportsTwoAST,
    originalImportsExtraAST,
    originalMarkerAST,
    originalSimpleAST;

  beforeAll(async () => {
    originalImportsAST = await parseFixture('with-imports.md');
    originalImportsTwoAST = await parseFixture('with-imports-two.md');
    originalImportsExtraAST = await parseFixture('with-imports-extra.md');
    originalSimpleAST = await parseFixture('simple.md');
    originalMarkerAST = await parseFixture('with-marker-imports.md');
  });

  beforeEach(() => {
    importsAST = cloneDeep(originalImportsAST);
    importsTwoAST = cloneDeep(originalImportsTwoAST);
    importsExtraAST = cloneDeep(originalImportsExtraAST);
    simpleAST = cloneDeep(originalSimpleAST);
    markerAST = cloneDeep(originalMarkerAST);
    correctFile = toVfile(
      path.resolve(__dirname, '../__fixtures__/with-imports.md')
    );
    incorrectFile = toVfile(
      path.resolve(__dirname, '../__fixtures__/incorrect-path/with-imports.md')
    );
  });

  it('should return a function', () => {
    expect.assertions(1);
    const plugin = addImports();

    expect(typeof plugin).toEqual('function');
  });

  it('should fail when the imported file is null', done => {
    const plugin = addImports();
    const next = err => {
      if (err) {
        done();
      } else {
        done('An error should have been thrown by addImports');
      }
    };
    plugin(importsAST, null, next);
  });

  it('should proceed when the imported file exists', done => {
    const plugin = addImports();
    plugin(importsAST, correctFile, done);
  });

  it('should fail when the imported file cannot be found', done => {
    expect.assertions(1);
    console.error = jest.fn();
    const plugin = addImports();

    // we have to rely on the next callback, because that is how you get error
    // messages out of transformers
    const next = err => {
      if (err) {
        expect(console.error).toHaveBeenCalledTimes(2);
        done();
      } else {
        done('An error should have been thrown by addImports');
      }
    };
    plugin(importsAST, incorrectFile, next);
  });

  it('should modify the tree when there are imports', done => {
    expect.assertions(1);
    const plugin = addImports();
    const next = err => {
      if (err) {
        done(err);
      } else {
        expect(importsAST).not.toEqual(originalImportsAST);
        done();
      }
    };
    plugin(importsAST, correctFile, next);
  });

  it('should NOT modify the tree when there are NO imports', done => {
    expect.assertions(1);
    const plugin = addImports();
    const next = err => {
      if (err) {
        done(err);
      } else {
        expect(simpleAST).toEqual(originalSimpleAST);
        done();
      }
    };
    plugin(simpleAST, correctFile, next);
  });

  it('should remove all import statements', done => {
    expect.assertions(2);
    const selector = 'leafDirective[name=import]';
    const plugin = addImports();
    const importNodes = selectAll(selector, importsAST);

    expect(importNodes.length).toBe(1);

    const next = err => {
      if (err) {
        done(err);
      } else {
        const importNodes = selectAll(selector, importsAST);
        expect(importNodes.length).toBe(0);
        done();
      }
    };
    plugin(importsAST, correctFile, next);
  });

  it('should not remove an ::import without the required attributes', done => {
    expect.assertions(2);
    const selector = 'leafDirective[name=import]';
    const plugin = addImports();
    const importNodes = selectAll(selector, importsExtraAST);

    expect(importNodes.length).toBe(3);

    const next = err => {
      if (err) {
        done(err);
      } else {
        const importNodes = selectAll(selector, importsExtraAST);
        expect(importNodes.length).toBe(1);
        done();
      }
    };
    plugin(importsExtraAST, correctFile, next);
  });

  it('should remove all matching ::use statements', done => {
    expect.assertions(2);
    const selector = 'leafDirective[name=use]';
    const plugin = addImports();
    const components = selectAll(selector, importsAST);

    // one matching component and two other jsx nodes
    expect(components.length).toBe(1);

    const next = err => {
      if (err) {
        done(err);
      } else {
        const components = selectAll(selector, importsAST);
        expect(components.length).toBe(0);
        done();
      }
    };
    plugin(importsAST, correctFile, next);
  });

  it('should replace the ::use statement with the imported content', done => {
    // checks the contents of script.md are there after the import step
    expect.assertions(2);
    const plugin = addImports();

    const next = err => {
      if (err) {
        done(err);
      } else {
        const jsNodes = selectAll('code[lang=js]', importsAST);
        expect(jsNodes.length).toBe(4);

        const codeValues = jsNodes.map(({ value }) => value);
        expect(codeValues).toEqual(
          expect.arrayContaining([
            `for (let index = 0; index < array.length; index++) {
  const element = array[index];
  // imported from script.md
}`
          ])
        );
        done();
      }
    };
    plugin(importsAST, correctFile, next);
  });

  it('should handle multiple import statements', done => {
    // checks the contents of script.md are there after the import step
    expect.assertions(4);
    const plugin = addImports();

    const next = err => {
      if (err) {
        done(err);
      } else {
        const jsNodes = selectAll('code[lang=js]', importsTwoAST);
        expect(jsNodes.length).toBe(4);

        const codeValues = jsNodes.map(({ value }) => value);
        expect(codeValues).toEqual(
          expect.arrayContaining([
            `for (let index = 0; index < array.length; index++) {
  const element = array[index];
  // imported from script.md
}`
          ])
        );
        const cssNodes = selectAll('code[lang=css]', importsTwoAST);
        expect(cssNodes.length).toBe(2);

        const cssValues = cssNodes.map(({ value }) => value);
        expect(cssValues).toEqual(
          expect.arrayContaining([
            `div {
  background: red
}`
          ])
        );
        done();
      }
    };
    plugin(importsTwoAST, correctFile, next);
  });

  it('should reject imported files with editable region markers', done => {
    expect.assertions(1);
    console.error = jest.fn();
    const plugin = addImports();
    const next = err => {
      if (err) {
        expect(console.error).toHaveBeenCalledTimes(2);
        done();
      } else {
        done('An error should have been thrown by addImports');
      }
    };
    plugin(markerAST, correctFile, next);
  });

  it('should have an output to match the snapshot', done => {
    const plugin = addImports();
    const next = err => {
      if (err) {
        done(err);
      } else {
        expect(importsAST).toMatchSnapshot();
        done();
      }
    };
    plugin(importsAST, correctFile, next);
  });
});
