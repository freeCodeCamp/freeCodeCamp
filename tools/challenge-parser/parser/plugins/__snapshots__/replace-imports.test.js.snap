// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`replace-imports should have an output to match the snapshot 1`] = `
{
  "children": [
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 18,
              "line": 3,
              "offset": 67,
            },
            "start": {
              "column": 3,
              "line": 3,
              "offset": 52,
            },
          },
          "type": "text",
          "value": "--description--",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 18,
          "line": 3,
          "offset": 67,
        },
        "start": {
          "column": 1,
          "line": 3,
          "offset": 50,
        },
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 5,
              "offset": 80,
            },
            "start": {
              "column": 1,
              "line": 5,
              "offset": 69,
            },
          },
          "type": "text",
          "value": "Paragraph 1",
        },
      ],
      "position": {
        "end": {
          "column": 12,
          "line": 5,
          "offset": 80,
        },
        "start": {
          "column": 1,
          "line": 5,
          "offset": 69,
        },
      },
      "type": "paragraph",
    },
    {
      "lang": "html",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 9,
          "offset": 106,
        },
        "start": {
          "column": 1,
          "line": 7,
          "offset": 82,
        },
      },
      "type": "code",
      "value": "code example",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 19,
              "line": 11,
              "offset": 126,
            },
            "start": {
              "column": 3,
              "line": 11,
              "offset": 110,
            },
          },
          "type": "text",
          "value": "--instructions--",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 19,
          "line": 11,
          "offset": 126,
        },
        "start": {
          "column": 1,
          "line": 11,
          "offset": 108,
        },
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 13,
              "offset": 139,
            },
            "start": {
              "column": 1,
              "line": 13,
              "offset": 128,
            },
          },
          "type": "text",
          "value": "Paragraph 0",
        },
      ],
      "position": {
        "end": {
          "column": 12,
          "line": 13,
          "offset": 139,
        },
        "start": {
          "column": 1,
          "line": 13,
          "offset": 128,
        },
      },
      "type": "paragraph",
    },
    {
      "lang": "html",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 17,
          "offset": 167,
        },
        "start": {
          "column": 1,
          "line": 15,
          "offset": 141,
        },
      },
      "type": "code",
      "value": "code example 0",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 12,
              "line": 19,
              "offset": 180,
            },
            "start": {
              "column": 3,
              "line": 19,
              "offset": 171,
            },
          },
          "type": "text",
          "value": "--hints--",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 12,
          "line": 19,
          "offset": 180,
        },
        "start": {
          "column": 1,
          "line": 19,
          "offset": 169,
        },
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 11,
              "line": 21,
              "offset": 192,
            },
            "start": {
              "column": 1,
              "line": 21,
              "offset": 182,
            },
          },
          "type": "text",
          "value": "First hint",
        },
      ],
      "position": {
        "end": {
          "column": 11,
          "line": 21,
          "offset": 192,
        },
        "start": {
          "column": 1,
          "line": 21,
          "offset": 182,
        },
      },
      "type": "paragraph",
    },
    {
      "lang": "js",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 25,
          "offset": 216,
        },
        "start": {
          "column": 1,
          "line": 23,
          "offset": 194,
        },
      },
      "type": "code",
      "value": "// test code",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 18,
              "line": 27,
              "offset": 235,
            },
            "start": {
              "column": 1,
              "line": 27,
              "offset": 218,
            },
          },
          "type": "text",
          "value": "Second hint with ",
        },
        {
          "position": {
            "end": {
              "column": 24,
              "line": 27,
              "offset": 241,
            },
            "start": {
              "column": 18,
              "line": 27,
              "offset": 235,
            },
          },
          "type": "html",
          "value": "<code>",
        },
        {
          "position": {
            "end": {
              "column": 28,
              "line": 27,
              "offset": 245,
            },
            "start": {
              "column": 24,
              "line": 27,
              "offset": 241,
            },
          },
          "type": "text",
          "value": "code",
        },
        {
          "position": {
            "end": {
              "column": 35,
              "line": 27,
              "offset": 252,
            },
            "start": {
              "column": 28,
              "line": 27,
              "offset": 245,
            },
          },
          "type": "html",
          "value": "</code>",
        },
      ],
      "position": {
        "end": {
          "column": 35,
          "line": 27,
          "offset": 252,
        },
        "start": {
          "column": 1,
          "line": 27,
          "offset": 218,
        },
      },
      "type": "paragraph",
    },
    {
      "lang": "js",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 31,
          "offset": 281,
        },
        "start": {
          "column": 1,
          "line": 29,
          "offset": 254,
        },
      },
      "type": "code",
      "value": "// more test code",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 11,
              "line": 34,
              "offset": 294,
            },
            "start": {
              "column": 3,
              "line": 34,
              "offset": 286,
            },
          },
          "type": "text",
          "value": "--seed--",
        },
      ],
      "depth": 1,
      "position": {
        "end": {
          "column": 11,
          "line": 34,
          "offset": 294,
        },
        "start": {
          "column": 1,
          "line": 34,
          "offset": 284,
        },
      },
      "type": "heading",
    },
    {
      "children": [
        {
          "position": {
            "end": {
              "column": 21,
              "line": 36,
              "offset": 316,
            },
            "start": {
              "column": 4,
              "line": 36,
              "offset": 299,
            },
          },
          "type": "text",
          "value": "--seed-contents--",
        },
      ],
      "depth": 2,
      "position": {
        "end": {
          "column": 21,
          "line": 36,
          "offset": 316,
        },
        "start": {
          "column": 1,
          "line": 36,
          "offset": 296,
        },
      },
      "type": "heading",
    },
    {
      "lang": "html",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 43,
          "offset": 363,
        },
        "start": {
          "column": 1,
          "line": 38,
          "offset": 318,
        },
      },
      "type": "code",
      "value": "<html>
  <body>
  </body>
</html>",
    },
    {
      "lang": "css",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 49,
          "offset": 405,
        },
        "start": {
          "column": 1,
          "line": 45,
          "offset": 365,
        },
      },
      "type": "code",
      "value": "body {
  background: green;
}",
    },
    {
      "attributes": {
        "id": "custom-name",
      },
      "children": [],
      "name": "id",
      "position": {
        "end": {
          "column": 19,
          "line": 51,
          "offset": 425,
        },
        "start": {
          "column": 1,
          "line": 51,
          "offset": 407,
        },
      },
      "type": "leafDirective",
    },
    {
      "lang": "js",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 55,
          "offset": 449,
        },
        "start": {
          "column": 1,
          "line": 53,
          "offset": 427,
        },
      },
      "type": "code",
      "value": "var x = 'y';",
    },
    {
      "lang": "js",
      "meta": null,
      "position": {
        "end": {
          "column": 4,
          "line": 6,
          "offset": 125,
        },
        "start": {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "code",
      "value": "for (let index = 0; index < array.length; index++) {
  const element = array[index];
  // imported from script.md
}",
    },
  ],
  "position": {
    "end": {
      "column": 1,
      "line": 58,
      "offset": 477,
    },
    "start": {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
